### ISSUE TITLE
Should Docker do better error reporting when the entrypoint has an invalid shebang?

### ISSUE THREAD

**Issue by jpetazzo** [docker/docker#29420]
**Description**

Specifying an invalid shebang will show a very terse error message.

Yes, this is a known issue (see #10668 and #10738 and perhaps #27780), but I would like to be that guy:
![image](https://cloud.githubusercontent.com/assets/171481/21197147/fbb76c86-c1ff-11e6-9c1d-1e795abda88a.png)
... i.e. I'd like to fight for the users! 🤘 

(Also, `bash` is handling this issue more gracefully than we are, and there's no way we gotta get no shame from a shell whose source code is not even on GitHub... no hard feelings, bash, you're still my favorite 🐚 , but I can't even easily link to a snippet of that wonderful code of yours!)

**Steps to reproduce the issue:**

```
mkdir shebang
echo '#!/bin/nonexistent' >shebang/shebang
chmod +x shebang/shebang
cat >shebang/Dockerfile <<EOF
FROM alpine
COPY shebang /
ENTRYPOINT ["/shebang"]
EOF
docker build -t shebang shebang
docker run --rm shebang
```

**Describe the results you received:**

```
standard_init_linux.go:175: exec user process caused "no such file or directory"
```

**Describe the results you expected:**

If I execute this entrypoint directly from the shell, I get instead:
```
$ shebang/shebang
-bash: shebang/shebang: /bin/nonexistent: bad interpreter: No such file or directory
```

... which is more helpful to track down the issue.

Of course, this is just `bash` being fancy and all. If you try to `os.system("shebang/shebang")` from a Python script, or even `ash -c shebang/shebang`, you will go back to a good, old, unhelpful, UNIX-unfriendly error message.

![image](https://cloud.githubusercontent.com/assets/171481/21197310/a39faf30-c200-11e6-9e5d-931439b7ff84.png)

**Additional information you deem important (e.g. issue happens only occasionally):**

I would like to thank [@anotherhowie](https://twitter.com/AnotherHowie/status/809097681729650689) for pointing out that we could do better.

Oh, and this issue is particularly prone to happen if you are using Windows, where line-breaks are different from Linux. You can _very easily_ end up with an invalid shebang, because it will be `#!/usr/bin/env python^M` and Linux will be like "y'all can take your caret em' and shove it where it ain't no sunshine, I'm not executing that stuff!" (Linux can be cranky at times.)

**Output of `docker version`:**

Tested with 1.13.0-rc3 and 1.12.1.

**Output of `docker info`:**

Whoa there, whoa there. We don't want to get into _that_ conversation, do we? I already added two silly images, you don't want me to paste pages and pages of irrelevant information, do you?

```
NOPE NOPE NOPE
```

... I thought so!

**Additional environment details (AWS, VirtualBox, physical, etc.):**

I don't know if we should try to be smart (like bash and zsh do) on the Engine side, or the client side. Client side would probably be good enough, but iono!

**Comment by thaJeztah** [comment]
ping @mlaventure, I think this error was done inside containerd or RunC? Think we can do better?

**Comment by thaJeztah** [comment]
> Tested with 1.13.0-rc3 and 1.12.1.

@jpetazzo pssst; [time to update](https://github.com/docker/docker/releases/tag/v1.12.4)

**Comment by mlaventure** [comment]
Yes, it comes straight from `runc` which doesn't include the name of the binary trying to be exec'ed.

I guess we could post process it in docker to add that information when get a `no such file or directory`

**Comment by justincormack** [comment]
I don't think you can get a better error message without actually parsing the shebang yourself, the kernel does not distinguish the file you are trying to execute not being there or bad shebang (or shared libraries missing) see man page "ENOENT The file filename or a script or ELF interpreter does not exist, or a shared library needed for the file or interpreter cannot be found.". You could check if the file exists when you get `ENOENT` and guess that it is a shebang/shared library issue.

**Comment by srizzling** [comment]
I was hurt by this recently, and would like to help, but have no idea where to start or any ideas of how to tackle this issue other than doing the above suggestion.

**Comment by adminy** [comment]
Not sure if this is related, probably is. I'm getting:
> OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: no such file or directory: unknown

And I'm not allowed to see what the running command was because GitHub Actions is pretty good at hiding stuff away.
problems like these are what make docker not such a great technology. Too abstract of an error message for me to do something about it.
